llmfoundry/models/mpt/modeling_mpt.py

```diff
@@ -448,6 +454,142 @@ def construct_blocks(self, config: MPTConfig) -> nn.ModuleList:
             ) for _ in range(config.n_layers)
         ])
 
+    def _construct_blocks_with_overrides(
+        self,
+        config: MPTConfig,
+        block_args: Dict[str, Any],
+    ) -> nn.ModuleList:
+        if config.block_overrides is None:
+            raise ValueError(
+                'config.block_overrides should not be None when calling _construct_blocks_with_overrides.',
+            )
+        modules_order_expanded = {}
+        for type in 'start', 'repeating_pattern', 'end':
+            modules_order_expanded[type] = []
+            if type in config.block_overrides:
+                for block in config.block_overrides[type]:
+                    if not isinstance(block['repeat'],
+                                      int) or block['repeat'] < 0:
+                        raise ValueError(
+                            'repeat should be a non-negative integer.',
+                        )
+                    modules_order_expanded[type].extend([block['name']] *
+                                                        block['repeat'])
+
+        start_len = len(modules_order_expanded['start'])
+        repeating_pattern_len = len(modules_order_expanded['repeating_pattern'])
+        end_len = len(modules_order_expanded['end'])
+
+        if repeating_pattern_len > 0:
+            if (
+                config.n_layers - (start_len + end_len)
+            ) % repeating_pattern_len != 0:
+                raise ValueError(
+                    'Number of layers should be divisible by the specified custom modules order.',
+                )
+            modules_order_expanded[
+                'repeating_pattern'
+            ] = modules_order_expanded['repeating_pattern'] * (
+                (config.n_layers -
+                 (start_len + end_len)) // repeating_pattern_len
+            )
+
+        model_modules_order_expanded = modules_order_expanded[
+            'start'] + modules_order_expanded['repeating_pattern'
+                                             ] + modules_order_expanded['end']
+
+        module_list = []
+        layer_description_list = []
+        if len(model_modules_order_expanded) != config.n_layers:
+            raise ValueError(
+                f'The specified block overrides do not match the number of layers: {len(model_modules_order_expanded)} vs {config.n_layers}.',
+            )
+
+        for i in range(config.n_layers):
```

Comment:
suggest helper function for this that just constructs the list of block args, and then construct the blocks after that whole list of configs has been created

Reply:
Done.